cmd/deploy/deploy.go

```diff
@@ -0,0 +1,469 @@
+				d.Status.NumberUnavailable == 0 {
+				log.Infof("Meshnet Healthy")
+				return nil
+			}
+		}
+	}
+}
+
+type Cluster interface {
+	Deploy(context.Context) error
+}
+
+type Ingress interface {
+	Deploy(context.Context) error
+	SetKClient(kubernetes.Interface)
+	Healthy(context.Context) error
+}
+
+type CNI interface {
+	Deploy(context.Context) error
+	SetKClient(kubernetes.Interface)
+	Healthy(context.Context) error
+}
+
+type Deployment struct {
+	Cluster Cluster
+	Ingress Ingress
+	CNI     CNI
+}
+
+type DeploymentConfig struct {
+	Cluster ClusterSpec `yaml:"cluster"`
+	Ingress IngressSpec `yaml:"ingress"`
+	CNI     CNISpec     `yaml:"cni"`
+}
+
+func NewDeployment(cfg *DeploymentConfig) (*Deployment, error) {
+	d := &Deployment{}
+	switch cfg.Cluster.Kind {
+	case "Kind":
+		log.Infof("Using kind scenario")
+		v := &KindSpec{}
+		if err := cfg.Cluster.Spec.Decode(v); err != nil {
+			return nil, err
+		}
+		d.Cluster = v
+	default:
+		return nil, fmt.Errorf("cluster type not supported: %s", cfg.Cluster.Kind)
+	}
+	switch cfg.CNI.Kind {
+	case "Meshnet":
+		v := &MeshnetSpec{}
+		if err := cfg.CNI.Spec.Decode(v); err != nil {
+			return nil, err
+		}
+		d.CNI = v
+	default:
+		return nil, fmt.Errorf("CNI type not supported: %s", cfg.CNI.Kind)
+	}
+	switch cfg.Ingress.Kind {
+	case "MetalLB":
+		v := &MetalLBSpec{}
+		if err := cfg.Ingress.Spec.Decode(v); err != nil {
+			return nil, err
+		}
+		d.Ingress = v
+	default:
+		return nil, fmt.Errorf("ingress type not supported: %s", cfg.Ingress.Kind)
+	}
+	return d, nil
+}
+
+var (
+	deploymentBasePath string
+)
+
+func deploymentFromArg(p string) (*DeploymentConfig, string, error) {
+	deploymentBasePath, err := filepath.Abs(p)
+	if err != nil {
+		return nil, "", err
+	}
+	b, err := ioutil.ReadFile(deploymentBasePath)
+	if err != nil {
+		return nil, "", err
+	}
+	deploymentBasePath = filepath.Dir(deploymentBasePath)
+	dCfg := &DeploymentConfig{}
+	decoder := yaml.NewDecoder(bytes.NewBuffer(b))
+	decoder.KnownFields(true)
+	if err := decoder.Decode(dCfg); err != nil {
+		return nil, "", err
+	}
+	return dCfg, deploymentBasePath, nil
+}
+
+func deployFn(cmd *cobra.Command, args []string) error {
+	if len(args) != 1 {
+		return fmt.Errorf("%s: missing args", cmd.Use)
+	}
+	dCfg, bp, err := deploymentFromArg(args[0])
+	if err != nil {
+		return err
+	}
+	deploymentBasePath = bp
+	d, err := NewDeployment(dCfg)
+	if err != nil {
+		return err
+	}
+	if err := d.Cluster.Deploy(cmd.Context()); err != nil {
+		return err
+	}
+	// Once cluster is up set kClient
+	kubecfg, err := cmd.Flags().GetString("kubecfg")
+	if err != nil {
+		return err
+	}
+	rCfg, err := clientcmd.BuildConfigFromFlags("", kubecfg)
+	if err != nil {
+		return err
+	}
+	kClient, err := kubernetes.NewForConfig(rCfg)
+	if err != nil {
+		return err
+	}
+	d.Ingress.SetKClient(kClient)
+	log.Infof("Validating cluster health")
+	if err := d.Ingress.Deploy(cmd.Context()); err != nil {
```

Comment:
In my experience it takes around 4 attempts to successfully deploy.

Maybe we should add the retry logic here

Reply:
hmm i have had zero failures so far with this setup

i don't want to guess at retrying things until we have data